mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java

```diff
@@ -118,38 +117,48 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
 
         assertRightsOnMailboxes(mailboxIds, mailboxSession, Right.Write);
 
-        Multimap<MailboxId, UpdatedFlags> updatedFlags = messageIdMapper.setFlags(messageId, mailboxIds, newState, replace);
-        for (Map.Entry<MailboxId, Collection<UpdatedFlags>> entry : updatedFlags.asMap().entrySet()) {
-            dispatchFlagsChange(mailboxSession, entry.getKey(), ImmutableList.copyOf(entry.getValue()));
-        }
+        messageIdMapper.setFlags(messageId, mailboxIds, newState, replace)
+            .flatMap(updatedFlags ->
+                Flux.fromIterable(updatedFlags.asMap().entrySet())
+                    .flatMap(Throwing.<Map.Entry<MailboxId, Collection<UpdatedFlags>>, Mono<Void>>function(entry ->
+                        dispatchFlagsChange(mailboxSession, entry.getKey(), ImmutableList.copyOf(entry.getValue())))
+                        .sneakyThrow())
+                    .then())
+            .block();
     }
 
     @Override
     public Set<MessageId> accessibleMessages(Collection<MessageId> messageIds, MailboxSession mailboxSession) throws MailboxException {
         MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
-        List<MailboxMessage> messageList = messageIdMapper.find(messageIds, MessageMapper.FetchType.Metadata);
-
-        ImmutableSet<MailboxId> allowedMailboxIds = getAllowedMailboxIds(mailboxSession, messageList, Right.Read);
-
-        return messageList.stream()
-            .filter(message -> allowedMailboxIds.contains(message.getMailboxId()))
+        return messageIdMapper.find(messageIds, MessageMapper.FetchType.Metadata)
+            .collect(Guavate.toImmutableList())
```

Comment:
I did an implementation of that with a groupedFlux: https://github.com/linagora/james-project/pull/3313

Feel free to use it if it helps :-)

Reply:
not really :/

Reply:
What do you mean?